File: lib/db/feeds.ts
import type { Feed } from "../types"
import { supabase } from "../supabase/client"
import { getCurrentUserId, toISOString } from "./core"
import { logger } from "../logger"

/**
 * Save multiple feeds to database
 * Upserts feeds with current user ownership
 */
export async function saveFeeds(feeds: Feed[]): Promise<{ success: boolean; error?: string }> {
  const userId = await getCurrentUserId()

  const dbRows = feeds.map(feed => ({
    id: feed.id,
    title: feed.title,
    url: feed.url,
    description: feed.description || null,
    category: feed.category || null,
    folder_id: feed.folderId || null,
    order: feed.order ?? 0,
    unread_count: feed.unreadCount ?? 0,
    refresh_interval: feed.refreshInterval ?? 60,
    user_id: userId,
    last_fetched: toISOString(feed.lastFetched),
  }))

  logger.debug({ userId, feedCount: feeds.length }, 'Saving feeds')
  const { data, error } = await supabase.from("feeds").upsert(dbRows).select()

  if (error) {
    logger.error({ error, userId, feedCount: feeds.length }, 'Failed to save feeds')
    if (error.code === '23505') {
      return { success: false, error: 'duplicate' }
    }
    throw error
  }

  logger.info({ userId, savedCount: data?.length || 0 }, 'Feeds saved successfully')
  return { success: true }
}

/**
 * Load all feeds for current user
 * Returns feeds ordered by order field
 */
export async function loadFeeds(): Promise<Feed[]> {
  logger.debug('Loading feeds')
  const { data, error } = await supabase
    .from("feeds")
    .select("*")
    .order("order", { ascending: true })

  if (error) {
    logger.error({ error }, 'Failed to load feeds')
    throw error
  }

  logger.debug({ feedCount: data?.length || 0 }, 'Feeds loaded successfully')
  return (data || []).map(row => ({
    id: row.id,
    title: row.title,
    url: row.url,
    description: row.description || undefined,
    category: row.category || undefined,
    folderId: row.folder_id || undefined,
    order: row.order,
    unreadCount: row.unread_count,
    refreshInterval: row.refresh_interval,
    lastFetched: row.last_fetched ? new Date(row.last_fetched) : undefined,
  }))
}

/**
 * Update a single feed
 * Allows partial updates of feed properties
 */
export async function updateFeed(feedId: string, updates: Partial<Feed>): Promise<{ success: boolean; error?: string }> {
  const userId = await getCurrentUserId()

  const updateData: any = {}

  if (updates.title !== undefined) updateData.title = updates.title
  if (updates.url !== undefined) updateData.url = updates.url
  if (updates.description !== undefined) updateData.description = updates.description || null
  if (updates.category !== undefined) updateData.category = updates.category || null
  if (updates.folderId !== undefined) updateData.folder_id = updates.folderId || null
  if (updates.order !== undefined) updateData.order = updates.order
  if (updates.unreadCount !== undefined) updateData.unread_count = updates.unreadCount
  if (updates.refreshInterval !== undefined) updateData.refresh_interval = updates.refreshInterval
  if (updates.lastFetched !== undefined) updateData.last_fetched = toISOString(updates.lastFetched)

  logger.debug({ feedId, userId, updateFields: Object.keys(updateData) }, 'Updating feed')

  const { error } = await supabase
    .from("feeds")
    .update(updateData)
    .eq("id", feedId)
    .eq("user_id", userId)

  if (error) {
    logger.error({ error, feedId, userId }, 'Failed to update feed')
    if (error.code === '23505') {
      return { success: false, error: 'duplicate' }
    }
    throw error
  }

  logger.info({ feedId, userId, updatedFields: Object.keys(updateData) }, 'Feed updated successfully')
  return { success: true }
}

/**
 * Delete a feed and all its articles
 * This will cascade delete due to foreign key constraints
 */
export async function deleteFeed(feedId: string): Promise<void> {
  const userId = await getCurrentUserId()
  logger.debug({ feedId, userId }, 'Deleting feed')

  const { error } = await supabase.from("feeds").delete().eq("id", feedId)

  if (error) {
    logger.error({ error, feedId, userId }, 'Failed to delete feed')
    throw error
  }

  logger.info({ feedId, userId }, 'Feed deleted successfully')
}

File: backend/app/services/db/feeds.py
"""
Feed database service using Supabase Python SDK.

Mirrors the functionality of lib/db/feeds.ts
"""

import logging
from typing import Optional, List

from .base import BaseDbService

logger = logging.getLogger(__name__)


def _cancel_feed_task(feed_id: str):
    """Cancel Celery refresh task for a feed (lazy import to avoid circular deps)."""
    try:
        from app.celery_app.tasks import cancel_feed_refresh
        cancel_feed_refresh(feed_id)
    except Exception as e:
        logger.warning(f"Failed to cancel feed task {feed_id}: {e}")


class FeedService(BaseDbService):
    """Service for feed database operations."""

    table_name = "feeds"

    # Fields allowed in update operations
    UPDATE_FIELDS = {
        "title", "url", "description", "category", "folder_id", "order",
        "unread_count", "refresh_interval", "last_fetched",
        "last_fetch_status", "last_fetch_error", "enable_deduplication",
    }

    def _row_to_dict(self, row: dict) -> dict:
        """Convert database row to feed dict."""
        return {
            "id": row["id"],
            "title": row["title"],
            "url": row["url"],
            "description": row.get("description"),
            "category": row.get("category"),
            "folder_id": row.get("folder_id"),
            "order": row["order"],
            "unread_count": row["unread_count"],
            "refresh_interval": row["refresh_interval"],
            "user_id": row["user_id"],
            "last_fetched": row.get("last_fetched"),
            "last_fetch_status": row.get("last_fetch_status"),
            "last_fetch_error": row.get("last_fetch_error"),
            "enable_deduplication": row.get("enable_deduplication", False),
            "created_at": row.get("created_at"),
        }

    def save_feeds(self, feeds: List[dict]) -> dict:
        """
        Save multiple feeds to database.
        Upserts feeds with current user ownership.

        Args:
            feeds: List of feed dictionaries

        Returns:
            dict with success status and optional error
        """
        db_rows = []
        for feed in feeds:
            db_rows.append(self._dict_to_row({
                "id": str(feed.get("id")) if feed.get("id") else None,
                "title": feed["title"],
                "url": feed["url"],
                "description": feed.get("description"),
                "category": feed.get("category"),
                "folder_id": str(feed["folder_id"]) if feed.get("folder_id") else None,
                "order": feed.get("order", 0),
                "unread_count": feed.get("unread_count", 0),
                "refresh_interval": feed.get("refresh_interval", 60),
                "last_fetched": feed.get("last_fetched"),
                "last_fetch_status": feed.get("last_fetch_status"),
                "last_fetch_error": feed.get("last_fetch_error"),
                "enable_deduplication": feed.get("enable_deduplication", False),
            }))

        logger.debug(f"Saving {len(feeds)} feeds", extra={'user_id': self.user_id})

        try:
            response = self._table().upsert(db_rows).execute()

            if response.data:
                logger.info(
                    f"Saved {len(response.data)} feeds",
                    extra={'user_id': self.user_id}
                )
                return {"success": True, "data": response.data}
            return {"success": True, "data": []}
        except Exception as e:
            logger.error(
                "Failed to save feeds",
                extra={'user_id': self.user_id, 'error': str(e)}
            )
            if self._is_duplicate_error(e):
                return {"success": False, "error": "duplicate"}
            raise

    def load_feeds(self) -> List[dict]:
        """
        Load all feeds for current user.
        Returns feeds ordered by order field.
        """
        logger.debug("Loading feeds", extra={'user_id': self.user_id})

        feeds = self._get_many(order_by="order", order_desc=False)

        logger.debug(f"Loaded {len(feeds)} feeds", extra={'user_id': self.user_id})
        return feeds

    def get_feed(self, feed_id: str) -> Optional[dict]:
        """Get a single feed by ID."""
        try:
            return self._get_one({"id": feed_id})
        except Exception as e:
            logger.error(
                f"Database error getting feed {feed_id}",
                extra={'user_id': self.user_id, 'error': str(e)}
            )
            return None

    def update_feed(self, feed_id: str, updates: dict) -> dict:
        """
        Update a single feed.
        Allows partial updates of feed properties.

        Args:
            feed_id: Feed UUID
            updates: Dictionary of fields to update

        Returns:
            dict with success status and optional error
        """
        logger.info(
            f"update_feed called: feed_id={feed_id}, updates={updates}",
            extra={'user_id': self.user_id}
        )

        update_data = self._prepare_update_data(updates, self.UPDATE_FIELDS)

        logger.info(
            f"Updating feed {feed_id} with data: {update_data}",
            extra={'user_id': self.user_id}
        )

        if not update_data:
            logger.warning(
                f"No valid update data for feed {feed_id}",
                extra={'user_id': self.user_id}
            )
            return {"success": True, "message": "No fields to update"}

        try:
            self._update_one(feed_id, update_data)

            logger.info(
                f"Updated feed {feed_id}",
                extra={'user_id': self.user_id}
            )
            return {"success": True}
        except Exception as e:
            logger.error(
                f"Failed to update feed {feed_id}",
                extra={'user_id': self.user_id, 'error': str(e)}
            )
            if self._is_duplicate_error(e):
                return {"success": False, "error": "duplicate"}
            raise

    def delete_feed(self, feed_id: str) -> dict:
        """
        Delete a feed and all its articles.
        Returns deletion statistics.

        Args:
            feed_id: Feed UUID

        Returns:
            dict with articles_deleted and feed_deleted
        """
        logger.debug(
            f"Starting feed deletion: {feed_id}",
            extra={'user_id': self.user_id}
        )

        # Step 0: Cancel scheduled Celery refresh task
        _cancel_feed_task(feed_id)

        # Step 1: Count and delete articles
        articles_response = (
            self.supabase.table("articles")
            .select("id")
            .eq("feed_id", feed_id)
            .eq("user_id", self.user_id)
            .execute()
        )

        article_count = len(articles_response.data or [])
        logger.debug(
            f"Found {article_count} articles to delete for feed {feed_id}",
            extra={'user_id': self.user_id}
        )

        articles_deleted = 0
        if article_count > 0:
            self.supabase.table("articles") \
                .delete() \
                .eq("feed_id", feed_id) \
                .eq("user_id", self.user_id) \
                .execute()
            articles_deleted = article_count
            logger.info(
                f"Deleted {articles_deleted} articles for feed {feed_id}",
                extra={'user_id': self.user_id}
            )

        # Step 2: Delete the feed
        self._delete_one(feed_id)

        logger.info(
            f"Deleted feed {feed_id} with {articles_deleted} articles",
            extra={'user_id': self.user_id}
        )
        return {"articles_deleted": articles_deleted, "feed_deleted": True}
